query/sources/impl/src/main/java/org/codice/ddf/admin/sources/csw/persist/SaveCswConfiguration.java

```diff
@@ -0,0 +1,110 @@
+import static org.codice.ddf.admin.sources.fields.type.SourceConfigUnionField.SOURCE_NAME_FIELD;
+
+import java.util.List;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang.StringUtils;
+import org.codice.ddf.admin.api.action.Message;
+import org.codice.ddf.admin.api.fields.Field;
+import org.codice.ddf.admin.api.fields.ObjectField;
+import org.codice.ddf.admin.common.actions.BaseAction;
+import org.codice.ddf.admin.configurator.ConfiguratorFactory;
+import org.codice.ddf.admin.sources.fields.ServicePid;
+import org.codice.ddf.admin.sources.fields.SourceInfoField;
+import org.codice.ddf.admin.sources.fields.type.CswSourceConfigurationField;
+
+import com.google.common.collect.ImmutableList;
+
+public class SaveCswConfiguration extends BaseAction<SourceInfoField> {
+
+    public static final String ID = "saveCswSource";
+
+    public static final String DESCRIPTION =
+            "Saves a CSW source configuration. If a pid is specified, the source configuration specified by the pid will be updated.";
+
+    private CswSourceConfigurationField config;
+
+    private ServicePid servicePid;
+
+    private ConfiguratorFactory configuratorFactory;
+
+
+    public SaveCswConfiguration(ConfiguratorFactory configuratorFactory) {
+        super(ID, DESCRIPTION, new SourceInfoField());
+        config = new CswSourceConfigurationField();
+        servicePid = new ServicePid();
+
+        config.isRequired(true);
+        setRequiredField(config, FACTORY_PID_FIELD, true);
+        setRequiredField(config, SOURCE_NAME_FIELD, true);
+        setRequiredField(config, ENDPOINT_URL_FIELD, true);
+
+        this.configuratorFactory = configuratorFactory;
+    }
+
+    @Override
+    public SourceInfoField performAction() {
+        if (StringUtils.isNotEmpty(servicePid.getValue())) {
+            return updateExistingConfig();
+        }
+
+        return persistNewConfig();
+    }
+
+    @Override
+    public List<Field> getArguments() {
+        return ImmutableList.of(config, servicePid);
+    }
+
+    private SourceInfoField updateExistingConfig() {
+        List<Message> results = updateConfig(servicePid, config, configuratorFactory,
+                cswConfigToServiceProps(config));
+
+        if(CollectionUtils.isNotEmpty(results)) {
+            results.forEach(this::addArgumentMessage);
+            return null;
+        }
+        return createSourceInfoField(ID, true, config);
+    }
+
+    private SourceInfoField persistNewConfig() {
+        List<Message> results = persist(config, configuratorFactory,
+                cswConfigToServiceProps(config));
+
+        if (CollectionUtils.isNotEmpty(results)) {
+            results.forEach(this::addArgumentMessage);
+            return null;
+        }
+        return createSourceInfoField(ID, true, config);
+    }
+
+    private void setRequiredField(ObjectField objectField, String fieldName, boolean required) {
```

Comment:
Delete this method, we should avoid matching by fieldName as much as possible since they can be changed.